metecho/api/serializers.py

```diff
@@ -356,15 +357,46 @@ def update(self, instance, validated_data):
                 org_type=SCRATCH_ORG_TYPES.Dev
             )
             for org in orgs:
-                org.queue_delete(originating_user_id=originating_user_id)
+                if new_user := self._valid_reassign(
+                    "dev", instance.assigned_dev, validated_data["assigned_dev"]
+                ):
+                    org.queue_reassign(
+                        new_user=new_user, originating_user_id=originating_user_id
+                    )
+                else:
+                    org.queue_delete(originating_user_id=originating_user_id)
         if instance.assigned_qa != validated_data["assigned_qa"]:
             orgs = instance.scratchorg_set.active().filter(
                 org_type=SCRATCH_ORG_TYPES.QA
             )
             for org in orgs:
-                org.queue_delete(originating_user_id=originating_user_id)
+                if new_user := self._valid_reassign(
+                    "qa", instance.assigned_qa, validated_data["assigned_qa"]
+                ):
+                    org.queue_reassign(
+                        new_user=new_user, originating_user_id=originating_user_id
+                    )
+                else:
+                    org.queue_delete(originating_user_id=originating_user_id)
         return super().update(instance, validated_data)
 
+    def _valid_reassign(self, type_, old_assignee, new_assignee):
+        old_user = self.get_matching_assigned_user(
+            type_, {f"assigned_{type_}": old_assignee}
+        )
+        new_user = self.get_matching_assigned_user(
+            type_, {f"assigned_{type_}": new_assignee}
+        )
+        if old_user and new_user and old_user.sf_username == new_user.sf_username:
+            return new_user
```

Comment:
This looks good to me for reassigning... The next step (re-using a soft-deleted org) will likely require storing either the creating `user` or the creating `user.sf_username` (likely the latter) on the org itself.

Reply:
Can you expand on this comment for me? I'm not sure I follow right now. Thanks!

Reply:
@wlonk This PR so far only addresses re-assigning a task-role-with-scratch-org from one user to another. The other part of this story is re-using a soft-deleted scratch org when assigning a task-role to a new user (from no-user-assigned state). To do that, I think we likely want to track the creating `sf_username` on the org itself. See the description on https://trello.com/c/hxJDjFuD/136-when-re-assigning-a-task-user-role-transfer-ownership-of-existing-scratch-org-instead-of-deleting-it

Does that makes sense? I'm happy to jump on a call to talk through it.

Reply:
OK, I think that what I have added will do this. Let me know if it works for you, and if it is good, I will sort out the `queue_delete` method, too.